src/rewrite.ml

```diff
@@ -0,0 +1,235 @@
+    match Solve.infer g_ctxt pred with
+    | Some a -> a
+    | None   -> fatal_no_pos "Cannot infer type of pred."
+  in
+  (*************************************************************************)
+
+  (* Construct the type of the new goal, which is obtained by performing the
+   *  subsitution by r, and adding the P symbol on top *)
+  (* FIXME: There is some duplication here, that could be avoided, by perhaps
+   * using Eval.snf or some other standard function. *)
+  let new_type = Bindlib.subst pred_bind r   in
+  let new_type = Appl(p_symb, new_type)      in
+
+  (*************************************************************************)
+  (* TODO: Remove this. We keep it for now to print types, for debugging. *)
+  let new_type_type =
+    match Solve.infer g_ctxt new_type with
+    | Some a -> a
+    | None   -> fatal_no_pos "Cannot infer type of the new type."
+  in
+  (*************************************************************************)
+
+  (* Build the new meta variable. As a term it will be used in eq_ind later on
+   * and as a meta it will be used to update the goal, once the rewrite tactic
+   * has finished. *)
+  let new_m_term = Ctxt.make_meta g_ctxt new_type  in
+  let new_m = match new_m_term with
+    | Meta(x, _) -> x
+    | _          -> fatal_no_pos "Should not get here."
+  in
+  (* Get the inductive principle associated with Leibniz equality to transform
+   * a proof of the new goal to a proof of the previous goal. *)
+
+  (* FIXME: When a Logic module with a notion of equality is defined get this
+  from that module. *)
+  let eq_ind        = Symb(Sign.find (Sign.current_sign()) "eqind")     in
+  (* Build the final lambda term that the tactic has produced. *)
+  let term_produced = add_args eq_ind [a ; l ; r ; t ; pred ; new_m_term]  in
+
+  (*************************************************************************)
+  (* TODO: Remove this. We keep it for now to print types, for debugging. *)
+  let term_produced_type =
+    match Solve.infer g_ctxt term_produced with
+    | Some a -> Eval.snf a
+    | None   -> fatal_no_pos "Cannot infer type of the term produced."
+  in
+  (*************************************************************************)
+
+  (* Type check the term used to instantiate the old meta. *)
+  if not (Solve.check g_ctxt term_produced g.g_type) then
+    fatal_no_pos "Fatal error: Wrong type.";
+  (* Update the value of the meta of the current goal. *)
+  let meta_env = Array.map Bindlib.unbox (Env.vars_of_env g.g_hyps)  in
+  let b = Bindlib.bind_mvar (to_tvars meta_env) (lift term_produced) in
+  let b = Bindlib.unbox b in
+  g_meta.meta_value := Some b ;
+  (*Since #REWRITE does not change g_hyps etc. we just update the necessary
+   * parameters in the original goal. *)
+  let thm =
+    {thm with t_goals = {g with g_meta = new_m ; g_type = new_type}::gs} in
+  theorem := Some thm ;
+
+  begin
+    print_endline " -------------- #REWRITE information -------------- " ;
```

Comment:
Use `wrn` and a newline here. Note that you should only use the functions of the `Console` module for outputs. You can also use the logging function defined at the top of `src/handle.ml`. If you do so, the output will only be enabled if you give the corresponding character to the `--debut` option.